### PR TITLE
Create the DatabaseDriver

### PR DESCRIPTION
**Pull request by jasonwyatt**
Follow-up PR will include tests to verify it plays nicely with the ReferenceModeStore.

### CODE REVIEW

File: java/arcs/core/crdt/CrdtSet.kt [hunk 1]
```diff
@@ -307,6 +318,24 @@ class CrdtSet<T : Referencable>(
...
+        if (this::class != other::class) return false
+
+        other as CrdtSet<*>
+
+        if (_data != other._data) return false
+
+        return true
+    }
+
+    override fun hashCode(): Int {
```
Comment (csilvestrini) [inline java/arcs/core/crdt/CrdtSet.kt]: Use `=` shorthand?

File: java/arcs/core/data/util/BUILD [hunk 2]
```diff
@@ -8,7 +8,7 @@ licenses(["notice"])
 package(default_visibility = ["//visibility:public"])
 
 arcs_kt_library(
-    name = "util",
+    name = "data-util",
```
Comment (csilvestrini) [inline java/arcs/core/data/util/BUILD]: Why this change?

File: java/arcs/core/storage/driver/Database.kt [hunk 3]
```diff
@@ -151,27 +165,194 @@ object DatabaseDriverProvider : DriverProvider {
...
 
+    @Suppress("UNCHECKED_CAST")
     override suspend fun registerReceiver(
         token: String?,
         receiver: suspend (data: Data, version: Int) -> Unit
     ) {
-        TODO("not implemented")
+        this.receiver = receiver
+        val (pendingReceiverData, pendingReceiverVersion) = localDataMutex.withLock {
+            var res = localData?.takeIf { this.token != token } to localVersion
```
Comment (csilvestrini) [inline java/arcs/core/storage/driver/Database.kt]: What does "res" stand for? Can you rename it?

File: java/arcs/core/data/util/BUILD (hunk 2 above)
Comment (jasonwyatt) [reply java/arcs/core/data/util/BUILD]: Apparently Kotlin/JS doesn't appreciate it when there are two build targets as deps with the same "name". Maybe we could instead modify the build macro to include the entire package path in the javascript lib filename?

Comment (csilvestrini) [reply java/arcs/core/data/util/BUILD]: Ooof, that sounds incredibly annoying. Might need to look into that.
